pkg/raft/testdata/lazy_replication.txt

```diff
@@ -0,0 +1,123 @@
+# leader's.
+stabilize
+----
+> 1 handling Ready
+  Ready MustSync=true:
+  Entries:
+  1/12 EntryNormal "data-1"
+  1/13 EntryNormal "data-2"
+  Messages:
+  1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "data-1"]
+  1->3 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "data-1"]
+  1->2 MsgApp Term:1 Log:1/12 Commit:11 Entries:[1/13 EntryNormal "data-2"]
+  1->3 MsgApp Term:1 Log:1/12 Commit:11 Entries:[1/13 EntryNormal "data-2"]
+> 2 receiving messages
+  1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "data-1"]
+  1->2 MsgApp Term:1 Log:1/12 Commit:11 Entries:[1/13 EntryNormal "data-2"]
+> 3 receiving messages
+  1->3 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "data-1"]
+  1->3 MsgApp Term:1 Log:1/12 Commit:11 Entries:[1/13 EntryNormal "data-2"]
+> 2 handling Ready
+  Ready MustSync=true:
+  Entries:
+  1/12 EntryNormal "data-1"
+  1/13 EntryNormal "data-2"
+  Messages:
+  2->1 MsgAppResp Term:1 Log:0/12 Commit:11
+  2->1 MsgAppResp Term:1 Log:0/13 Commit:11
+> 3 handling Ready
+  Ready MustSync=true:
+  Entries:
+  1/12 EntryNormal "data-1"
+  1/13 EntryNormal "data-2"
+  Messages:
+  3->1 MsgAppResp Term:1 Log:0/12 Commit:11
+  3->1 MsgAppResp Term:1 Log:0/13 Commit:11
+> 1 receiving messages
+  2->1 MsgAppResp Term:1 Log:0/12 Commit:11
+  2->1 MsgAppResp Term:1 Log:0/13 Commit:11
+  3->1 MsgAppResp Term:1 Log:0/12 Commit:11
+  3->1 MsgAppResp Term:1 Log:0/13 Commit:11
+> 1 handling Ready
+  Ready MustSync=true:
+  HardState Term:1 Vote:1 Commit:13 Lead:1 LeadEpoch:1
+  CommittedEntries:
+  1/12 EntryNormal "data-1"
+  1/13 EntryNormal "data-2"
+  Messages:
+  1->2 MsgApp Term:1 Log:1/13 Commit:12
+  1->3 MsgApp Term:1 Log:1/13 Commit:12
+  1->2 MsgApp Term:1 Log:1/13 Commit:13
+  1->3 MsgApp Term:1 Log:1/13 Commit:13
+> 2 receiving messages
+  1->2 MsgApp Term:1 Log:1/13 Commit:12
+  1->2 MsgApp Term:1 Log:1/13 Commit:13
+> 3 receiving messages
+  1->3 MsgApp Term:1 Log:1/13 Commit:12
+  1->3 MsgApp Term:1 Log:1/13 Commit:13
+> 2 handling Ready
+  Ready MustSync=true:
+  HardState Term:1 Vote:1 Commit:13 Lead:1 LeadEpoch:1
+  CommittedEntries:
+  1/12 EntryNormal "data-1"
+  1/13 EntryNormal "data-2"
+  Messages:
+  2->1 MsgAppResp Term:1 Log:0/13 Commit:12
+  2->1 MsgAppResp Term:1 Log:0/13 Commit:13
+> 3 handling Ready
+  Ready MustSync=true:
+  HardState Term:1 Vote:1 Commit:13 Lead:1 LeadEpoch:1
+  CommittedEntries:
+  1/12 EntryNormal "data-1"
+  1/13 EntryNormal "data-2"
+  Messages:
+  3->1 MsgAppResp Term:1 Log:0/13 Commit:12
+  3->1 MsgAppResp Term:1 Log:0/13 Commit:13
+> 1 receiving messages
+  2->1 MsgAppResp Term:1 Log:0/13 Commit:12
+  2->1 MsgAppResp Term:1 Log:0/13 Commit:13
+  3->1 MsgAppResp Term:1 Log:0/13 Commit:12
+  3->1 MsgAppResp Term:1 Log:0/13 Commit:13
+
+# The leader has converged to a fully replicated state.
+status 1
+----
+1: StateReplicate match=13 next=14 sentCommit=11 matchCommit=11
+2: StateReplicate match=13 next=14 sentCommit=13 matchCommit=13
```

Comment:
Yes, I tripped on this too. We discussed this effect with @iskettaneh a bunch while he was working on #129943, and agreed that the leader's commit indices in `Progress` struct is ok to be stale. The leader does not use its own indices for replication (there are no explicit self-directed `MsgApp`s on the leader that would depend on them). Some background:

`matchCommit` is a commit index known to be durably persisted on this peer (we use this fact to quiesce `MsgApp` pings when the peer is fully up-to-date). On the leader, we don't know if the commit index was synced, and don't enforce it. If we enforced it (for consistency, in order to make this `matchCommit` converge), it could cause unnecessary syncs in some cases.

We could update `sentCommit` to be equal to `raftLog.commit` at least. But since `matchCommit` isn't converging, we would similarly be "confused" when looking at this line. The best we would be able to achieve here is that `sentCommit == raftLog.commit`, and `matchCommit` is lagging by one `MsgStorageAppend` roundtrip.